# US Gun Violence

## Preliminary analysis

In [1]:
# importing libraries

import pandas as pd
import numpy as np
import requests
from bs4 import BeautifulSoup
import re

In [2]:
# data set main location

loc = r'D:\DA_Projects\Gun Violence USA\Data Set'

In [3]:
# get the list of data files from location

import os

for i in os.walk(loc,topdown = False):
    files = i[2]


In [4]:
# list of files

files

['2014.csv',
 '2015.csv',
 '2016.csv',
 '2017.csv',
 '2018.csv',
 '2019.csv',
 '2020,2021,2022,2023.csv',
 '2020.csv',
 '2021.csv',
 '2022.csv',
 'Mother Jones - Mass Shootings Database, 1982 - 2023 - Sheet1.csv',
 'School Shooting - 2021,2022, 2023.csv']

In [5]:
print(files)

['2014.csv', '2015.csv', '2016.csv', '2017.csv', '2018.csv', '2019.csv', '2020,2021,2022,2023.csv', '2020.csv', '2021.csv', '2022.csv', 'Mother Jones - Mass Shootings Database, 1982 - 2023 - Sheet1.csv', 'School Shooting - 2021,2022, 2023.csv']


* Individual Year Data - '2014.csv', '2015.csv', '2016.csv', '2017.csv', '2018.csv', '2019.csv', '2020.csv', '2021.csv', '2022.csv'
* Combined data file for 4 years - '2020,2021,2022,2023.csv'
* Specific to school shooting - School Shooting - 2021,2022, 2023.csv
* Data from other source - 'Mother Jones - Mass Shootings Database, 1982 - 2023 - Sheet1.csv'

In [6]:
# Create individual paths for reading the .csv files as and when required

filepaths = []

for i in files:
    filepaths.append(os.path.join(loc,i))

In [7]:
filepaths

['D:\\DA_Projects\\Gun Violence USA\\Data Set\\2014.csv',
 'D:\\DA_Projects\\Gun Violence USA\\Data Set\\2015.csv',
 'D:\\DA_Projects\\Gun Violence USA\\Data Set\\2016.csv',
 'D:\\DA_Projects\\Gun Violence USA\\Data Set\\2017.csv',
 'D:\\DA_Projects\\Gun Violence USA\\Data Set\\2018.csv',
 'D:\\DA_Projects\\Gun Violence USA\\Data Set\\2019.csv',
 'D:\\DA_Projects\\Gun Violence USA\\Data Set\\2020,2021,2022,2023.csv',
 'D:\\DA_Projects\\Gun Violence USA\\Data Set\\2020.csv',
 'D:\\DA_Projects\\Gun Violence USA\\Data Set\\2021.csv',
 'D:\\DA_Projects\\Gun Violence USA\\Data Set\\2022.csv',
 'D:\\DA_Projects\\Gun Violence USA\\Data Set\\Mother Jones - Mass Shootings Database, 1982 - 2023 - Sheet1.csv',
 'D:\\DA_Projects\\Gun Violence USA\\Data Set\\School Shooting - 2021,2022, 2023.csv']

In [8]:
# checking 2014 data

data_2014 = pd.read_csv(filepaths[1])

In [9]:
data_2014

Incident ID      Incident Date         State City Or County   
0         476321  December 31, 2015     Louisiana    New Orleans  \
1         473882  December 27, 2015     Tennessee        Jackson   
2         472268  December 26, 2015  Pennsylvania   Philadelphia   
3         471867  December 25, 2015       Florida   Jacksonville   
4         471860  December 25, 2015       Alabama         Mobile   
..           ...                ...           ...            ...   
331       274867    January 6, 2015       Florida          Miami   
332       273535    January 4, 2015      Virginia        Roanoke   
333       273397    January 4, 2015         Texas         Dallas   
334       273965    January 2, 2015       Georgia       Savannah   
335       272016    January 1, 2015     Tennessee        Memphis   

                          Address  # Killed  # Injured  Operations  
0     1900 block of Amelia Street         0          6         NaN  
1                   North Parkway         0          4         NaN  
2              4210 Macalester St         0          4         NaN  
3             Franklin and Odessa         0          4         NaN  
4            785 Schillinger Rd S         0          4         NaN  
..                            ...       ...        ...         ...  
331  1300 block of NW 62nd Street         1          3         NaN  
332        3634 Shenandoah Ave NW         2          4         NaN  
333     2000 block of Ben Hur St.         3          1         NaN  
334   500 block of W. 54th Street         1          4         NaN  
335          I-240 and Poplar Ave         0          5         NaN  

[336 rows x 8 columns]

In [10]:
# getting basic info about columns of 2014 data
# since we are going to combine the data for all the years
# also we are going to cross check the data with already combined data for any errors.

data_2014.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 336 entries, 0 to 335
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Incident ID     336 non-null    int64  
 1   Incident Date   336 non-null    object 
 2   State           336 non-null    object 
 3   City Or County  336 non-null    object 
 4   Address         335 non-null    object 
 5   # Killed        336 non-null    int64  
 6   # Injured       336 non-null    int64  
 7   Operations      0 non-null      float64
dtypes: float64(1), int64(3), object(4)
memory usage: 21.1+ KB


In [11]:
data_2014.columns

Index(['Incident ID', 'Incident Date', 'State', 'City Or County', 'Address',
       '# Killed', '# Injured', 'Operations'],
      dtype='object')

- 336 rows
- we see that there are 8 columns ( this has to be verified with all other data sets )
    - Incident ID 
        - int
        - not mandatory, may be needed to check duplicates
        - no null
    - Incident Date
        - data type - object - should be of datatime or date data type
        - no null
        - mandatory - required
    - State
        - object - should be string
        - mandatory
        - no- null
    - City Or County
        - object - should be string
        - mandatory
        - no- null
    - Address
        - object - should be string
        - mandatory
        - no- null
    - #Killed
        - int, okay - check for negetive, 0
    - #Injured
        - int, okay - check for negetive, 0        
    - Operations
        - object - column not needed - remove

Similarly all the years data and columns to be analysed

* lets check 2014 data completely

In [12]:
data_2014['# Killed'] < 3

0       True
1       True
2       True
3       True
4       True
       ...  
331     True
332     True
333    False
334     True
335     True
Name: # Killed, Length: 336, dtype: bool

In [13]:
# check negative values
# `.values` to convert to numpy array and apply any() across rows

(data_2014.select_dtypes('int64').values < 0).any()

False

In [14]:
# no negative values found
# we can do the same for all the above data to verify
# lets check all the data

In [15]:
# lets verfiy the shapes of all the data sets to ensure they have same no. of columns

data_2014.shape

(336, 8)

In [16]:
columns_check = []
rows_check = []
headers = []

for i in filepaths[0:10]:
    df = pd.read_csv(i)
    columns_check.append(df.shape[1])
    rows_check.append(df.shape[0])
    headers.extend(df.columns)

In [17]:
columns_check

# all the datasets have same no. of columns, now lets analyse each dataset for null values

[8, 8, 8, 8, 8, 8, 8, 8, 8, 8]

In [18]:
rows_check

# row values differ with respect to the year, since each file represents data for different year

[272, 336, 383, 348, 336, 415, 2000, 610, 690, 646]

In [19]:
# lets check if the column headers are same

unique_headers = set(headers)

In [20]:
unique_headers

{'# Injured',
 '# Killed',
 'Address',
 'City Or County',
 'Incident Date',
 'Incident ID',
 'Operations',
 'State'}

In [21]:
len(unique_headers)

8

* The headers are good and have the same headers for all the data sets

* We have to check if individual year files and combined file of 2020,2021,2022 have the same number of rows.

In [22]:
individual_year_file_rows = 0

for i in rows_check[-3:]:
    individual_year_file_rows = individual_year_file_rows + i

In [23]:
individual_year_file_rows

1946

In [24]:
# lets check the combined data file of 2020, 2021, 2022

data_combined = pd.read_csv(filepaths[6])
data_combined

Incident ID   Incident Date           State City Or County   
0         2584808  April 30, 2023       Louisiana      Lafayette  \
1         2585253  April 30, 2023         Alabama     Birmingham   
2         2584772  April 30, 2023        Kentucky        Paducah   
3         2584779  April 30, 2023        Oklahoma  Oklahoma City   
4         2585718  April 30, 2023      California         Mojave   
...           ...             ...             ...            ...   
1995      1688753    May 26, 2020  South Carolina         Bowman   
1996      1687999    May 25, 2020       Louisiana   Lake Charles   
1997      1687219    May 24, 2020        Missouri    Saint Louis   
1998      1687037    May 24, 2020  South Carolina   Myrtle Beach   
1999      1687414    May 24, 2020         Alabama       Danville   

                                      Address  # Killed  # Injured  Operations  
0           W Congress Ave and University Ave         1          5         NaN  
1                      800 block of 48th St N         0          4         NaN  
2                                505 S 8th St         0          4         NaN  
3                          3000 W Memorial Rd         1          3         NaN  
4                         15900 block of H St         4          0         NaN  
...                                       ...       ...        ...         ...  
1995                    Dibble St and Rail St         3          5         NaN  
1996                               6780 US-90         0          6         NaN  
1997  Union Boulevard and Cote Brilliante Ave         0          4         NaN  
1998                12th Ave N and Ocean Blvd         0          4         NaN  
1999                    68 Flint Creek Pvt Dr         3          1         NaN  

[2000 rows x 8 columns]

* the above data consists of 2023 data as well
* count the 2023 rows and negate it from the total data to find the 2020 - 2022 rows

In [25]:
# total rows of the combined dataset

data_combined.shape[0]

2000

In [26]:
# 2023 data rows of the combined dataset

rows_2023 = data_combined[data_combined["Incident Date"].str.contains("2023")].shape[0]
rows_2023

185

In [27]:
#  2020 - 2022 data rows 

data_combined.shape[0] - rows_2023

1815

In [28]:
# individual year files have higher no. of rows for 2020 - 2022 than that of the combined data file of 2020 - 2022

print(individual_year_file_rows)
print(data_combined.shape[0] - rows_2023)

1946
1815


* Checking the rows difference for 2020, 2021, 2022 b/w individual and combined file

In [29]:
#combined file data 

rows_2022 = data_combined[data_combined["Incident Date"].str.contains("2022")].shape[0]
rows_2021 = data_combined[data_combined["Incident Date"].str.contains("2021")].shape[0]
rows_2020 = data_combined[data_combined["Incident Date"].str.contains("2020")].shape[0]

print(rows_2020, rows_2021,rows_2022)

479 690 646


In [30]:
# individual file row data for 2020, 2021, 2022

rows_check[-3:]

[610, 690, 646]

* its clear that 2020 has less no. of rows in the combined file that the individual file of year 2020
* 2023 data from combined year data file can be used

* Union all the individual year data
* To do a union, we have to check if the columns of all the data files are in correct order/sequece

In [31]:
headers_check = []

for i in filepaths[0:10]:
    df = pd.read_csv(i)
    headers_check.append(df.columns)

In [32]:
pd.DataFrame(headers_check)

0              1      2               3        4         5   
0  Incident ID  Incident Date  State  City Or County  Address  # Killed  \
1  Incident ID  Incident Date  State  City Or County  Address  # Killed   
2  Incident ID  Incident Date  State  City Or County  Address  # Killed   
3  Incident ID  Incident Date  State  City Or County  Address  # Killed   
4  Incident ID  Incident Date  State  City Or County  Address  # Killed   
5  Incident ID  Incident Date  State  City Or County  Address  # Killed   
6  Incident ID  Incident Date  State  City Or County  Address  # Killed   
7  Incident ID  Incident Date  State  City Or County  Address  # Killed   
8  Incident ID  Incident Date  State  City Or County  Address  # Killed   
9  Incident ID  Incident Date  State  City Or County  Address  # Killed   

           6           7  
0  # Injured  Operations  
1  # Injured  Operations  
2  # Injured  Operations  
3  # Injured  Operations  
4  # Injured  Operations  
5  # Injured  Operations  
6  # Injured  Operations  
7  # Injured  Operations  
8  # Injured  Operations  
9  # Injured  Operations

## Combining into a single data set - 2014 to 2023

* From the above it is clear that the columns are in order, and the data in files can be combined
* Combining Data using concat() function

In [33]:
# combining data till 2022

df0 = pd.DataFrame([],columns = data_2014.columns)

for i in filepaths[0:6]:
    df = pd.read_csv(i)
    df0 = pd.concat([df0,df], ignore_index = True)
    
for i in filepaths[7:10]:
    df = pd.read_csv(i)
    df0 = pd.concat([df0,df], ignore_index = True)

In [34]:
# 2014 to 2022 data

df0

Incident ID      Incident Date           State  City Or County   
0         271363  December 29, 2014       Louisiana     New Orleans  \
1         269679  December 27, 2014      California     Los Angeles   
2         270036  December 27, 2014      California      Sacramento   
3         269167  December 26, 2014        Illinois  East St. Louis   
4         268598  December 24, 2014        Missouri     Saint Louis   
...          ...                ...             ...             ...   
4031     2201641    January 1, 2022       Wisconsin         Kenosha   
4032     2202001    January 1, 2022        Illinois          Peoria   
4033     2201876    January 1, 2022  South Carolina          Dillon   
4034     2201848    January 1, 2022         Indiana      South Bend   
4035     2201677    January 1, 2022        Colorado          Denver   

                                  Address # Killed # Injured Operations  
0                     Poydras and Bolivar        0         4        NaN  
1     8800 block of South Figueroa Street        1         3        NaN  
2                4000 block of May Street        0         4        NaN  
3             2500 block of Summit Avenue        1         3        NaN  
4                           18th and Pine        1         3        NaN  
...                                   ...      ...       ...        ...  
4031                         1708 52nd St        1         3        NaN  
4032            2000 block of SW Adams St        2         2        NaN  
4033                           3261 Hwy 9        0         5        NaN  
4034             1500 block of S Miami St        0         4        NaN  
4035                        1919 Blake St        2         2        NaN  

[4036 rows x 8 columns]

In [35]:
# 2023 data

df_2023 = data_combined[data_combined["Incident Date"].str.contains("2023")]
df_2023

Incident ID    Incident Date           State City Or County   
0        2584808   April 30, 2023       Louisiana      Lafayette  \
1        2585253   April 30, 2023         Alabama     Birmingham   
2        2584772   April 30, 2023        Kentucky        Paducah   
3        2584779   April 30, 2023        Oklahoma  Oklahoma City   
4        2585718   April 30, 2023      California         Mojave   
..           ...              ...             ...            ...   
180      2492601  January 1, 2023        Illinois        Chicago   
181      2497601  January 1, 2023         Florida  Miami Gardens   
182      2492611  January 1, 2023  North Carolina         Durham   
183      2493102  January 1, 2023    Pennsylvania      Allentown   
184      2492314  January 1, 2023            Ohio       Columbus   

                               Address  # Killed  # Injured  Operations  
0    W Congress Ave and University Ave         1          5         NaN  
1               800 block of 48th St N         0          4         NaN  
2                         505 S 8th St         0          4         NaN  
3                   3000 W Memorial Rd         1          3         NaN  
4                  15900 block of H St         4          0         NaN  
..                                 ...       ...        ...         ...  
180             300 block of E 57th St         1          3         NaN  
181        NW 171st St and NW 30th Ave         0          9         NaN  
182                  1000 N Miami Blvd         0          5         NaN  
183                    1140 E Clair St         0          4         NaN  
184                  2830 Johnstown Rd         1          4         NaN  

[185 rows x 8 columns]

In [36]:
df0 = pd.concat([df0,df_2023], ignore_index = True)

In [37]:
# Combined data from 2014 - 2023

df0

Incident ID      Incident Date           State  City Or County   
0         271363  December 29, 2014       Louisiana     New Orleans  \
1         269679  December 27, 2014      California     Los Angeles   
2         270036  December 27, 2014      California      Sacramento   
3         269167  December 26, 2014        Illinois  East St. Louis   
4         268598  December 24, 2014        Missouri     Saint Louis   
...          ...                ...             ...             ...   
4216     2492601    January 1, 2023        Illinois         Chicago   
4217     2497601    January 1, 2023         Florida   Miami Gardens   
4218     2492611    January 1, 2023  North Carolina          Durham   
4219     2493102    January 1, 2023    Pennsylvania       Allentown   
4220     2492314    January 1, 2023            Ohio        Columbus   

                                  Address # Killed # Injured Operations  
0                     Poydras and Bolivar        0         4        NaN  
1     8800 block of South Figueroa Street        1         3        NaN  
2                4000 block of May Street        0         4        NaN  
3             2500 block of Summit Avenue        1         3        NaN  
4                           18th and Pine        1         3        NaN  
...                                   ...      ...       ...        ...  
4216               300 block of E 57th St        1         3        NaN  
4217          NW 171st St and NW 30th Ave        0         9        NaN  
4218                    1000 N Miami Blvd        0         5        NaN  
4219                      1140 E Clair St        0         4        NaN  
4220                    2830 Johnstown Rd        1         4        NaN  

[4221 rows x 8 columns]

## Analysing the combined data

In [38]:
df0.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4221 entries, 0 to 4220
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Incident ID     4221 non-null   object
 1   Incident Date   4221 non-null   object
 2   State           4221 non-null   object
 3   City Or County  4221 non-null   object
 4   Address         4209 non-null   object
 5   # Killed        4221 non-null   object
 6   # Injured       4221 non-null   object
 7   Operations      0 non-null      object
dtypes: object(8)
memory usage: 263.9+ KB


* The **Incident ID** which was of data type int64 is now of type object. - This has to be addressed
* **Incident Date** to datatime
* **Address** - check non-null
* **# Killed* - to int64
* **# Injured** - to int64
* Operations - delete


In [39]:
#creating a copy 

us_gunviolence = df0.copy()

In [40]:
us_gunviolence

Incident ID      Incident Date           State  City Or County   
0         271363  December 29, 2014       Louisiana     New Orleans  \
1         269679  December 27, 2014      California     Los Angeles   
2         270036  December 27, 2014      California      Sacramento   
3         269167  December 26, 2014        Illinois  East St. Louis   
4         268598  December 24, 2014        Missouri     Saint Louis   
...          ...                ...             ...             ...   
4216     2492601    January 1, 2023        Illinois         Chicago   
4217     2497601    January 1, 2023         Florida   Miami Gardens   
4218     2492611    January 1, 2023  North Carolina          Durham   
4219     2493102    January 1, 2023    Pennsylvania       Allentown   
4220     2492314    January 1, 2023            Ohio        Columbus   

                                  Address # Killed # Injured Operations  
0                     Poydras and Bolivar        0         4        NaN  
1     8800 block of South Figueroa Street        1         3        NaN  
2                4000 block of May Street        0         4        NaN  
3             2500 block of Summit Avenue        1         3        NaN  
4                           18th and Pine        1         3        NaN  
...                                   ...      ...       ...        ...  
4216               300 block of E 57th St        1         3        NaN  
4217          NW 171st St and NW 30th Ave        0         9        NaN  
4218                    1000 N Miami Blvd        0         5        NaN  
4219                      1140 E Clair St        0         4        NaN  
4220                    2830 Johnstown Rd        1         4        NaN  

[4221 rows x 8 columns]

## Cleaning

In [41]:
# deleting/removing "Operations" columns

us_gunviolence.drop(columns = ['Operations'], inplace = True)

In [42]:
us_gunviolence

Incident ID      Incident Date           State  City Or County   
0         271363  December 29, 2014       Louisiana     New Orleans  \
1         269679  December 27, 2014      California     Los Angeles   
2         270036  December 27, 2014      California      Sacramento   
3         269167  December 26, 2014        Illinois  East St. Louis   
4         268598  December 24, 2014        Missouri     Saint Louis   
...          ...                ...             ...             ...   
4216     2492601    January 1, 2023        Illinois         Chicago   
4217     2497601    January 1, 2023         Florida   Miami Gardens   
4218     2492611    January 1, 2023  North Carolina          Durham   
4219     2493102    January 1, 2023    Pennsylvania       Allentown   
4220     2492314    January 1, 2023            Ohio        Columbus   

                                  Address # Killed # Injured  
0                     Poydras and Bolivar        0         4  
1     8800 block of South Figueroa Street        1         3  
2                4000 block of May Street        0         4  
3             2500 block of Summit Avenue        1         3  
4                           18th and Pine        1         3  
...                                   ...      ...       ...  
4216               300 block of E 57th St        1         3  
4217          NW 171st St and NW 30th Ave        0         9  
4218                    1000 N Miami Blvd        0         5  
4219                      1140 E Clair St        0         4  
4220                    2830 Johnstown Rd        1         4  

[4221 rows x 7 columns]

In [43]:
us_gunviolence = us_gunviolence.astype({'Incident ID':'int64', '# Killed':'int64','# Injured':'int64'}, copy = False)

In [44]:
us_gunviolence["Incident ID"]

0        271363
1        269679
2        270036
3        269167
4        268598
         ...   
4216    2492601
4217    2497601
4218    2492611
4219    2493102
4220    2492314
Name: Incident ID, Length: 4221, dtype: int64

In [45]:
us_gunviolence.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4221 entries, 0 to 4220
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Incident ID     4221 non-null   int64 
 1   Incident Date   4221 non-null   object
 2   State           4221 non-null   object
 3   City Or County  4221 non-null   object
 4   Address         4209 non-null   object
 5   # Killed        4221 non-null   int64 
 6   # Injured       4221 non-null   int64 
dtypes: int64(3), object(4)
memory usage: 231.0+ KB


In [46]:
us_gunviolence[us_gunviolence['Address'].isna()]

Incident ID       Incident Date           State   City Or County   
365        411888   September 8, 2015         Georgia           Berlin  \
1427      1206476   September 9, 2018         Florida     Jacksonville   
2438      1725735        July 4, 2020  South Carolina         Barnwell   
2913      2115336   September 5, 2021        Missouri            Wyatt   
2968      2089557     August 10, 2021     Mississippi       Clarksdale   
3501      2438723    October 15, 2022     Mississippi        Itta Bena   
3541      2424602  September 23, 2022      California      Castroville   
3561      2412289  September 11, 2022        Michigan     Grand Rapids   
3683      2361849       July 17, 2022        Michigan          Detroit   
3774      2360288       June 11, 2022      New Mexico       Alamogordo   
3839      2354757        May 13, 2022       Louisiana    Napoleonville   
3860      2294515         May 3, 2022          Kansas  Cowley (county)   

     Address  # Killed  # Injured  
365      NaN         0          4  
1427     NaN         0          4  
2438     NaN         0          5  
2913     NaN         0          4  
2968     NaN         2          2  
3501     NaN         2          7  
3541     NaN         0          4  
3561     NaN         0          4  
3683     NaN         0          4  
3774     NaN         0          5  
3839     NaN         0          5  
3860     NaN         0          4

* Above are the missing adresses of the incidents
* We will keep them as we need the count of incidents happened

## Checking/Analysing County Data

In [47]:
# downlaoded the list of cities/towns by zipcode for each county - source opendatasoft

loc = r'C:\Users\pryns\Downloads\georef-united-states-of-america-zc-point@public (1).csv'

In [48]:
counties_list = pd.read_csv(loc,delimiter = ';')

In [162]:
counties_list.head()

Zip Code Official USPS city name Official USPS State Code   
0     90005             Los Angeles                       CA  \
1     90006             Los Angeles                       CA   
2     90008             Los Angeles                       CA   
3     90040             Los Angeles                       CA   
4     90045             Los Angeles                       CA   

  Official State Name  ZCTA  ZCTA parent  Population  Density   
0          California  True          NaN     39732.0  14151.8  \
1          California  True          NaN     59576.0  11981.7   
2          California  True          NaN     31754.0   3337.9   
3          California  True          NaN     12328.0    852.8   
4          California  True          NaN     40567.0   1459.6   

   Primary Official County Code Primary Official County Name   
0                          6037                  Los Angeles  \
1                          6037                  Los Angeles   
2                          6037                  Los Angeles   
3                          6037                  Los Angeles   
4                          6037                  Los Angeles   

     County Weights Official County Name Official County Code  Imprecise   
0  {"06037": "100"}          Los Angeles                 6037      False  \
1  {"06037": "100"}          Los Angeles                 6037      False   
2  {"06037": "100"}          Los Angeles                 6037      False   
3  {"06037": "100"}          Los Angeles                 6037      False   
4  {"06037": "100"}          Los Angeles                 6037      False   

   Military             Timezone             Geo Point  
0     False  America/Los_Angeles  34.05912, -118.30654  
1     False  America/Los_Angeles  34.04801, -118.29418  
2     False  America/Los_Angeles  34.00956, -118.34706  
3     False  America/Los_Angeles  33.99353, -118.14907  
4     False  America/Los_Angeles  33.95297, -118.40014

* above data has lots of information we need only the state, city and its county information
* creating a new dataframe below for the required information

In [49]:
counties_list_new = counties_list[['Official USPS city name', 'Official USPS State Code','Official State Name','Primary Official County Name']].copy()

In [50]:
counties_list_new.drop_duplicates(inplace = True)

In [163]:
counties_list_new.head()

Official USPS city name Official USPS State Code Official State Name   
0             Los Angeles                       CA          California  \
5            Santa Monica                       CA          California   
6                Torrance                       CA          California   
7                  Lomita                       CA          California   
8              Long Beach                       CA          California   

  Primary Official County Name  
0                  Los Angeles  
5                  Los Angeles  
6                  Los Angeles  
7                  Los Angeles  
8                  Los Angeles

* getting the unique list of state and city or county data from the original gun violence data to match and get the correct county information

In [52]:
search_counties = us_gunviolence[['State','City Or County']].drop_duplicates()
search_counties.head()

State  City Or County
0   Louisiana     New Orleans
1  California     Los Angeles
2  California      Sacramento
3    Illinois  East St. Louis
4    Missouri     Saint Louis

In [53]:
search_counties['City'] = search_counties['City Or County'].replace({r'\([\s\S]*\)':''}, regex=True)
search_counties['City'] = search_counties['City'].apply(lambda x : x.rstrip())

In [54]:
search_counties

State      City Or County            City
0      Louisiana         New Orleans     New Orleans
1     California         Los Angeles     Los Angeles
2     California          Sacramento      Sacramento
3       Illinois      East St. Louis  East St. Louis
4       Missouri         Saint Louis     Saint Louis
...          ...                 ...             ...
4188  California       Monterey Park   Monterey Park
4190  California              Goshen          Goshen
4192     Florida         Fort Pierce     Fort Pierce
4210        Utah  Cedar City (Enoch)      Cedar City
4212    Virginia            Dumfries        Dumfries

[1163 rows x 3 columns]

* merging the counties information with the gunviolence information of the counties

In [55]:
search_counties_new = search_counties.merge(right = counties_list_new, how = 'outer', left_on = ['City','State'], right_on = ['Official USPS city name','Official State Name'])

In [56]:
search_counties_new = search_counties_new[search_counties_new['City'].notna()].copy()

In [165]:
search_counties_new.head()

State          City Or County         City Official USPS city name   
0  Louisiana             New Orleans  New Orleans             New Orleans  \
1  Louisiana             New Orleans  New Orleans             New Orleans   
2  Louisiana  New Orleans (Gentilly)  New Orleans             New Orleans   
3  Louisiana  New Orleans (Gentilly)  New Orleans             New Orleans   
4  Louisiana   New Orleans (Algiers)  New Orleans             New Orleans   

  Official USPS State Code Official State Name Primary Official County Name  
0                       LA           Louisiana                      Orleans  
1                       LA           Louisiana                    Jefferson  
2                       LA           Louisiana                      Orleans  
3                       LA           Louisiana                    Jefferson  
4                       LA           Louisiana                      Orleans

*  checking for null values and duplicates within the merged data

In [57]:
search_counties_new_toupdate = search_counties_new[search_counties_new['Official USPS city name'].isna()].copy()

In [58]:
search_counties_new_toupdate

State                City Or County             City   
11          Illinois                East St. Louis   East St. Louis  \
76          New York                        Queens           Queens   
91             Texas                   Flour Bluff      Flour Bluff   
109       California                East Palo Alto   East Palo Alto   
148       Washington                        Skyway           Skyway   
162         Illinois                   Centreville      Centreville   
166         New York                     Manhattan        Manhattan   
170         Maryland                 Seat Pleasant    Seat Pleasant   
193        Tennessee                Lookout Valley   Lookout Valley   
349          Florida                    Allapattah       Allapattah   
355         Maryland                    Springdale       Springdale   
360         Missouri                       Kinloch          Kinloch   
368         New York                         Gates            Gates   
387         Missouri                        Tyrone           Tyrone   
506          Florida                    Plantation       Plantation   
536      Mississippi              Mc Cool (Mccool)          Mc Cool   
677          Alabama                       Brandon          Brandon   
743          Florida                   Brownsville      Brownsville   
1002  South Carolina              Mc Coll (Mccoll)          Mc Coll   
1078  South Carolina                     Mccormick        Mccormick   
1094    Pennsylvania  Mc Kees Rocks (Mckees Rocks)    Mc Kees Rocks   
1104        Illinois               Washington Park  Washington Park   
1127           Texas                     Mc Gregor        Mc Gregor   
1195          Kansas               Cowley (county)           Cowley   
1205      California                   Willowbrook      Willowbrook   
1253      California                        Goshen           Goshen   

     Official USPS city name Official USPS State Code Official State Name   
11                       NaN                      NaN                 NaN  \
76                       NaN                      NaN                 NaN   
91                       NaN                      NaN                 NaN   
109                      NaN                      NaN                 NaN   
148                      NaN                      NaN                 NaN   
162                      NaN                      NaN                 NaN   
166                      NaN                      NaN                 NaN   
170                      NaN                      NaN                 NaN   
193                      NaN                      NaN                 NaN   
349                      NaN                      NaN                 NaN   
355                      NaN                      NaN                 NaN   
360                      NaN                      NaN                 NaN   
368                      NaN                      NaN                 NaN   
387                      NaN                      NaN                 NaN   
506                      NaN                      NaN                 NaN   
536                      NaN                      NaN                 NaN   
677                      NaN                      NaN                 NaN   
743                      NaN                      NaN                 NaN   
1002                     NaN                      NaN                 NaN   
1078                     NaN                      NaN                 NaN   
1094                     NaN                      NaN                 NaN   
1104                     NaN                      NaN                 NaN   
1127                     NaN                      NaN                 NaN   
1195                     NaN                      NaN                 NaN   
1205                     NaN                      NaN                 NaN   
1253                     NaN                      NaN                 NaN   

     Primary Offi

* There are a lot of null values and a few corrections for city names

In [59]:
search_counties_new_toupdate.at[536,'City'] = 'McCool'
search_counties_new_toupdate.loc[536,'City']

'McCool'

In [60]:
search_counties_new_toupdate.at[1002,'City'] = 'McColl'
search_counties_new_toupdate.loc[1002,'City']

'McColl'

In [61]:
search_counties_new_toupdate.at[1094,'City'] = 'McKees Rocks'
search_counties_new_toupdate.loc[1094,'City']

'McKees Rocks'

In [62]:
search_counties_new_toupdate.at[1127,'City'] = 'McGregor'
search_counties_new_toupdate.loc[1127,'City']

'McGregor'

In [63]:
search_counties_new_toupdate.at[1078,'City'] = 'McCormick'
search_counties_new_toupdate.loc[1078,'City']

'McCormick'

In [64]:
search_counties_new_toupdate

State                City Or County             City   
11          Illinois                East St. Louis   East St. Louis  \
76          New York                        Queens           Queens   
91             Texas                   Flour Bluff      Flour Bluff   
109       California                East Palo Alto   East Palo Alto   
148       Washington                        Skyway           Skyway   
162         Illinois                   Centreville      Centreville   
166         New York                     Manhattan        Manhattan   
170         Maryland                 Seat Pleasant    Seat Pleasant   
193        Tennessee                Lookout Valley   Lookout Valley   
349          Florida                    Allapattah       Allapattah   
355         Maryland                    Springdale       Springdale   
360         Missouri                       Kinloch          Kinloch   
368         New York                         Gates            Gates   
387         Missouri                        Tyrone           Tyrone   
506          Florida                    Plantation       Plantation   
536      Mississippi              Mc Cool (Mccool)           McCool   
677          Alabama                       Brandon          Brandon   
743          Florida                   Brownsville      Brownsville   
1002  South Carolina              Mc Coll (Mccoll)           McColl   
1078  South Carolina                     Mccormick        McCormick   
1094    Pennsylvania  Mc Kees Rocks (Mckees Rocks)     McKees Rocks   
1104        Illinois               Washington Park  Washington Park   
1127           Texas                     Mc Gregor         McGregor   
1195          Kansas               Cowley (county)           Cowley   
1205      California                   Willowbrook      Willowbrook   
1253      California                        Goshen           Goshen   

     Official USPS city name Official USPS State Code Official State Name   
11                       NaN                      NaN                 NaN  \
76                       NaN                      NaN                 NaN   
91                       NaN                      NaN                 NaN   
109                      NaN                      NaN                 NaN   
148                      NaN                      NaN                 NaN   
162                      NaN                      NaN                 NaN   
166                      NaN                      NaN                 NaN   
170                      NaN                      NaN                 NaN   
193                      NaN                      NaN                 NaN   
349                      NaN                      NaN                 NaN   
355                      NaN                      NaN                 NaN   
360                      NaN                      NaN                 NaN   
368                      NaN                      NaN                 NaN   
387                      NaN                      NaN                 NaN   
506                      NaN                      NaN                 NaN   
536                      NaN                      NaN                 NaN   
677                      NaN                      NaN                 NaN   
743                      NaN                      NaN                 NaN   
1002                     NaN                      NaN                 NaN   
1078                     NaN                      NaN                 NaN   
1094                     NaN                      NaN                 NaN   
1104                     NaN                      NaN                 NaN   
1127                     NaN                      NaN                 NaN   
1195                     NaN                      NaN                 NaN   
1205                     NaN                      NaN                 NaN   
1253                     NaN                      NaN                 NaN   

     Primary Offi

In [65]:
search_counties_new_toupdate['Index'] = search_counties_new_toupdate.index

In [66]:
search_counties_new_toupdate.head()

State  City Or County            City Official USPS city name   
11     Illinois  East St. Louis  East St. Louis                     NaN  \
76     New York          Queens          Queens                     NaN   
91        Texas     Flour Bluff     Flour Bluff                     NaN   
109  California  East Palo Alto  East Palo Alto                     NaN   
148  Washington          Skyway          Skyway                     NaN   

    Official USPS State Code Official State Name Primary Official County Name   
11                       NaN                 NaN                          NaN  \
76                       NaN                 NaN                          NaN   
91                       NaN                 NaN                          NaN   
109                      NaN                 NaN                          NaN   
148                      NaN                 NaN                          NaN   

     Index  
11      11  
76      76  
91      91  
109    109  
148    148

In [67]:
search_counties_new_toupdate.drop(columns = ['Official USPS city name','Official USPS State Code','Official State Name','Primary Official County Name'],inplace = True)

In [68]:
search_counties_new_toupdate

State                City Or County             City  Index
11          Illinois                East St. Louis   East St. Louis     11
76          New York                        Queens           Queens     76
91             Texas                   Flour Bluff      Flour Bluff     91
109       California                East Palo Alto   East Palo Alto    109
148       Washington                        Skyway           Skyway    148
162         Illinois                   Centreville      Centreville    162
166         New York                     Manhattan        Manhattan    166
170         Maryland                 Seat Pleasant    Seat Pleasant    170
193        Tennessee                Lookout Valley   Lookout Valley    193
349          Florida                    Allapattah       Allapattah    349
355         Maryland                    Springdale       Springdale    355
360         Missouri                       Kinloch          Kinloch    360
368         New York                         Gates            Gates    368
387         Missouri                        Tyrone           Tyrone    387
506          Florida                    Plantation       Plantation    506
536      Mississippi              Mc Cool (Mccool)           McCool    536
677          Alabama                       Brandon          Brandon    677
743          Florida                   Brownsville      Brownsville    743
1002  South Carolina              Mc Coll (Mccoll)           McColl   1002
1078  South Carolina                     Mccormick        McCormick   1078
1094    Pennsylvania  Mc Kees Rocks (Mckees Rocks)     McKees Rocks   1094
1104        Illinois               Washington Park  Washington Park   1104
1127           Texas                     Mc Gregor         McGregor   1127
1195          Kansas               Cowley (county)           Cowley   1195
1205      California                   Willowbrook      Willowbrook   1205
1253      California                        Goshen           Goshen   1253

In [69]:
# County data extracted with the help of CHAT GPT for the missing city or county

df_update = pd.read_csv(r'C:\Users\pryns\Desktop\New Text Document3.csv',delimiter = ',')

In [70]:
df_update

City           State           County
0    East St. Louis        Illinois        St. Clair
1            Queens        New York           Queens
2       Flour Bluff           Texas           Nueces
3    East Palo Alto      California        San Mateo
4            Skyway      Washington             King
5       Centreville        Illinois        St. Clair
6         Manhattan        New York         New York
7     Seat Pleasant        Maryland  Prince George's
8    Lookout Valley       Tennessee         Hamilton
9        Allapattah         Florida       Miami-Dade
10       Springdale        Maryland  Prince George's
11          Kinloch        Missouri      Saint Louis
12            Gates        New York           Monroe
13           Tyrone        Missouri            Texas
14       Plantation         Florida          Broward
15           McCool     Mississippi           Attala
16          Brandon         Alabama           Shelby
17      Brownsville         Florida       Santa Rosa
18           McColl  South Carolina         Marlboro
19        McCormick  South Carolina        McCormick
20     McKees Rocks    Pennsylvania        Allegheny
21  Washington Park        Illinois        St. Clair
22         McGregor           Texas         McLennan
23           Cowley          Kansas           Cowley
24      Willowbrook      California      Los Angeles
25           Goshen      California           Tulare

* using the data to update the missing county information using merge

In [71]:
search_counties_new_toupdate = search_counties_new_toupdate.merge(right = df_update, how = 'outer', left_on = ['City','State'], right_on = ['City','State'],copy = False)

In [72]:
search_counties_new_toupdate

State                City Or County             City  Index   
0         Illinois                East St. Louis   East St. Louis     11  \
1         New York                        Queens           Queens     76   
2            Texas                   Flour Bluff      Flour Bluff     91   
3       California                East Palo Alto   East Palo Alto    109   
4       Washington                        Skyway           Skyway    148   
5         Illinois                   Centreville      Centreville    162   
6         New York                     Manhattan        Manhattan    166   
7         Maryland                 Seat Pleasant    Seat Pleasant    170   
8        Tennessee                Lookout Valley   Lookout Valley    193   
9          Florida                    Allapattah       Allapattah    349   
10        Maryland                    Springdale       Springdale    355   
11        Missouri                       Kinloch          Kinloch    360   
12        New York                         Gates            Gates    368   
13        Missouri                        Tyrone           Tyrone    387   
14         Florida                    Plantation       Plantation    506   
15     Mississippi              Mc Cool (Mccool)           McCool    536   
16         Alabama                       Brandon          Brandon    677   
17         Florida                   Brownsville      Brownsville    743   
18  South Carolina              Mc Coll (Mccoll)           McColl   1002   
19  South Carolina                     Mccormick        McCormick   1078   
20    Pennsylvania  Mc Kees Rocks (Mckees Rocks)     McKees Rocks   1094   
21        Illinois               Washington Park  Washington Park   1104   
22           Texas                     Mc Gregor         McGregor   1127   
23          Kansas               Cowley (county)           Cowley   1195   
24      California                   Willowbrook      Willowbrook   1205   
25      California                        Goshen           Goshen   1253   

             County  
0         St. Clair  
1            Queens  
2            Nueces  
3         San Mateo  
4              King  
5         St. Clair  
6          New York  
7   Prince George's  
8          Hamilton  
9        Miami-Dade  
10  Prince George's  
11      Saint Louis  
12           Monroe  
13            Texas  
14          Broward  
15           Attala  
16           Shelby  
17       Santa Rosa  
18         Marlboro  
19        McCormick  
20        Allegheny  
21        St. Clair  
22         McLennan  
23           Cowley  
24      Los Angeles  
25           Tulare

* Correcting City names in the original dataset

In [73]:
search_counties_new.at[536,'City'] = 'McCool'
search_counties_new.loc[536,'City']

'McCool'

In [74]:
search_counties_new.at[1002,'City'] = 'McColl'
search_counties_new.loc[1002,'City']

'McColl'

In [75]:
search_counties_new.at[1094,'City'] = 'McKees Rocks'
search_counties_new.loc[1094,'City']

'McKees Rocks'

In [76]:
search_counties_new.at[1127,'City'] = 'McGregor'
search_counties_new.loc[1127,'City']

'McGregor'

In [77]:
search_counties_new.at[1078,'City'] = 'McCormick'
search_counties_new.loc[1078,'City']

'McCormick'

In [78]:
search_counties_new[search_counties_new['City'].notna()].loc[1078]

State                           South Carolina
City Or County                       Mccormick
City                                 McCormick
Official USPS city name                    NaN
Official USPS State Code                   NaN
Official State Name                        NaN
Primary Official County Name               NaN
Name: 1078, dtype: object

* Updating the missing county names

In [79]:
search_counties_new_toupdate[['Index','County']].values

array([[11, 'St. Clair'],
       [76, 'Queens'],
       [91, 'Nueces'],
       [109, 'San Mateo'],
       [148, 'King'],
       [162, 'St. Clair'],
       [166, 'New York'],
       [170, "Prince George's"],
       [193, 'Hamilton'],
       [349, 'Miami-Dade'],
       [355, "Prince George's"],
       [360, 'Saint Louis'],
       [368, 'Monroe'],
       [387, 'Texas'],
       [506, 'Broward'],
       [536, 'Attala'],
       [677, 'Shelby'],
       [743, 'Santa Rosa'],
       [1002, 'Marlboro'],
       [1078, 'McCormick'],
       [1094, 'Allegheny'],
       [1104, 'St. Clair'],
       [1127, 'McLennan'],
       [1195, 'Cowley'],
       [1205, 'Los Angeles'],
       [1253, 'Tulare']], dtype=object)

In [80]:
for i in search_counties_new_toupdate[['Index','County']].values:
    search_counties_new.at[i[0],'Primary Official County Name'] = i[1]

In [81]:
search_counties_new[search_counties_new['City'].notna()].loc[1078]

State                           South Carolina
City Or County                       Mccormick
City                                 McCormick
Official USPS city name                    NaN
Official USPS State Code                   NaN
Official State Name                        NaN
Primary Official County Name         McCormick
Name: 1078, dtype: object

In [82]:
search_counties_new.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1257 entries, 0 to 1256
Data columns (total 7 columns):
 #   Column                        Non-Null Count  Dtype 
---  ------                        --------------  ----- 
 0   State                         1257 non-null   object
 1   City Or County                1257 non-null   object
 2   City                          1257 non-null   object
 3   Official USPS city name       1231 non-null   object
 4   Official USPS State Code      1231 non-null   object
 5   Official State Name           1231 non-null   object
 6   Primary Official County Name  1257 non-null   object
dtypes: object(7)
memory usage: 110.9+ KB


* Removing duplicates

In [104]:
search_counties_new[search_counties_new[['State','City Or County']].duplicated()]['City'].index

Index([   1,    3,    5,   26,   37,   38,   45,   46,   48,   49,   53,   56,
         57,   58,   68,   70,   86,   94,   95,  107,  122,  124,  126,  128,
        137,  138,  151,  156,  159,  161,  190,  196,  197,  199,  200,  210,
        215,  216,  229,  236,  252,  254,  257,  262,  281,  283,  286,  287,
        292,  316,  326,  336,  345,  363,  371,  372,  376,  380,  381,  410,
        412,  414,  528,  529,  531,  532,  574,  580,  587,  592,  609,  610,
        611,  641,  653,  666,  671,  692,  693,  725,  729,  785,  799,  907,
        908,  915,  948,  964,  970, 1072, 1107, 1113, 1163, 1190],
      dtype='int64')

In [103]:
search_counties_new[search_counties_new[['State','City Or County']].duplicated()][search_counties_new[search_counties_new[['State','City Or County']].duplicated()]['City'] == 'Kissimmee']

State City Or County       City Official USPS city name   
1190  Florida      Kissimmee  Kissimmee               Kissimmee  \

     Official USPS State Code Official State Name Primary Official County Name  
1190                       FL             Florida                         Polk

In [102]:
search_counties_new[search_counties_new['City'] == 'Kissimmee']

State City Or County       City Official USPS city name   
1189  Florida      Kissimmee  Kissimmee               Kissimmee  \
1190  Florida      Kissimmee  Kissimmee               Kissimmee   

     Official USPS State Code Official State Name Primary Official County Name  
1189                       FL             Florida                      Osceola  
1190                       FL             Florida                         Polk

In [106]:
search_counties_new_1 = search_counties_new.drop(index = search_counties_new[search_counties_new[['State','City Or County']].duplicated()]['City'].index).copy()

In [110]:
search_counties_new_1.drop(columns = ['Official USPS city name','Official State Name'], inplace = True)

In [121]:
search_counties_new_1.rename(columns = {'Official USPS State Code':'State Code','Primary Official County Name':'County'}, inplace = True)

In [123]:
search_counties_new_1[search_counties_new_1['State Code'].isna()]

State                City Or County             City   
11          Illinois                East St. Louis   East St. Louis  \
76          New York                        Queens           Queens   
91             Texas                   Flour Bluff      Flour Bluff   
109       California                East Palo Alto   East Palo Alto   
148       Washington                        Skyway           Skyway   
162         Illinois                   Centreville      Centreville   
166         New York                     Manhattan        Manhattan   
170         Maryland                 Seat Pleasant    Seat Pleasant   
193        Tennessee                Lookout Valley   Lookout Valley   
349          Florida                    Allapattah       Allapattah   
355         Maryland                    Springdale       Springdale   
360         Missouri                       Kinloch          Kinloch   
368         New York                         Gates            Gates   
387         Missouri                        Tyrone           Tyrone   
506          Florida                    Plantation       Plantation   
536      Mississippi              Mc Cool (Mccool)           McCool   
677          Alabama                       Brandon          Brandon   
743          Florida                   Brownsville      Brownsville   
1002  South Carolina              Mc Coll (Mccoll)           McColl   
1078  South Carolina                     Mccormick        McCormick   
1094    Pennsylvania  Mc Kees Rocks (Mckees Rocks)     McKees Rocks   
1104        Illinois               Washington Park  Washington Park   
1127           Texas                     Mc Gregor         McGregor   
1195          Kansas               Cowley (county)           Cowley   
1205      California                   Willowbrook      Willowbrook   
1253      California                        Goshen           Goshen   

     State Code           County  
11          NaN        St. Clair  
76          NaN           Queens  
91          NaN           Nueces  
109         NaN        San Mateo  
148         NaN             King  
162         NaN        St. Clair  
166         NaN         New York  
170         NaN  Prince George's  
193         NaN         Hamilton  
349         NaN       Miami-Dade  
355         NaN  Prince George's  
360         NaN      Saint Louis  
368         NaN           Monroe  
387         NaN            Texas  
506         NaN          Broward  
536         NaN           Attala  
677         NaN           Shelby  
743         NaN       Santa Rosa  
1002        NaN         Marlboro  
1078        NaN        McCormick  
1094        NaN        Allegheny  
1104        NaN        St. Clair  
1127        NaN         McLennan  
1195        NaN           Cowley  
1205        NaN      Los Angeles  
1253        NaN           Tulare

* Mapping missing STATE CODE information

In [125]:
us_state_code = counties_list_new[['Official USPS State Code','Official State Name']].drop_duplicates().copy()

In [126]:
us_state_code

Official USPS State Code       Official State Name
0                           CA                California
63                          HI                    Hawaii
64                          GU                      Guam
65                          OR                    Oregon
77                          WA                Washington
84                          SD              South Dakota
93                          PR               Puerto Rico
96                          MA             Massachusetts
114                         RI              Rhode Island
120                         NH             New Hampshire
133                         ME                     Maine
145                         VT                   Vermont
151                         CT               Connecticut
165                         NJ                New Jersey
186                         NY                  New York
244                         PA              Pennsylvania
308                         DE                  Delaware
310                         DC      District of Columbia
311                         VA                  Virginia
316                         MD                  Maryland
361                         WV             West Virginia
379                         NC            North Carolina
406                         SC            South Carolina
411                         GA                   Georgia
431                         FL                   Florida
465                         AL                   Alabama
493                         TN                 Tennessee
515                         MS               Mississippi
530                         KY                  Kentucky
574                         OH                      Ohio
618                         IN                   Indiana
649                         MI                  Michigan
678                         IA                      Iowa
713                         WI                 Wisconsin
751                         MN                 Minnesota
796                         ND              North Dakota
806                         MT                   Montana
815                         IL                  Illinois
858                         MO                  Missouri
887                         KS                    Kansas
914                         NE                  Nebraska
936                         LA                 Louisiana
946                         AR                  Arkansas
963                         OK                  Oklahoma
988                         TX                     Texas
1058                        CO                  Colorado
1068                        WY                   Wyoming
1073                        ID                     Idaho
1084                        UT                      Utah
1095                        AZ                   Arizona
1106                        NM                New Mexico
1122                        NV                    Nevada
1143                        AK                    Alaska
1574                        MP  Northern Mariana Islands
3204                        VI            Virgin Islands
17673                       AS            American Samoa

In [129]:
search_counties_new_1.merge(right = us_state_code, how = 'inner', left_on = ['State'], right_on = ['Official State Name'])

State          City Or County               City State Code   
0         Louisiana             New Orleans        New Orleans         LA  \
1         Louisiana  New Orleans (Gentilly)        New Orleans         LA   
2         Louisiana   New Orleans (Algiers)        New Orleans         LA   
3         Louisiana       Saint Martinville  Saint Martinville         LA   
4         Louisiana                Tallulah           Tallulah         LA   
...             ...                     ...                ...        ...   
1158        Wyoming                Cheyenne           Cheyenne         WY   
1159          Idaho                Caldwell           Caldwell         ID   
1160          Idaho                   Boise              Boise         ID   
1161  New Hampshire              Manchester         Manchester         NH   
1162         Hawaii                 Waianae            Waianae         HI   

            County Official USPS State Code Official State Name  
0          Orleans                       LA           Louisiana  
1          Orleans                       LA           Louisiana  
2          Orleans                       LA           Louisiana  
3       St. Martin                       LA           Louisiana  
4          Madison                       LA           Louisiana  
...            ...                      ...                 ...  
1158       Laramie                       WY             Wyoming  
1159        Canyon                       ID               Idaho  
1160           Ada                       ID               Idaho  
1161  Hillsborough                       NH       New Hampshire  
1162      Honolulu                       HI              Hawaii  

[1163 rows x 7 columns]

In [131]:
search_counties_new_1 = search_counties_new_1.merge(right = us_state_code, how = 'inner', left_on = ['State'], right_on = ['Official State Name'])

In [133]:
search_counties_new_1.drop(columns = ['State Code', 'Official State Name'], inplace = True)

In [135]:
search_counties_new_1.rename(columns = {'Official USPS State Code':'State Code'}, inplace = True)

In [136]:
search_counties_new_1

State          City Or County               City        County   
0         Louisiana             New Orleans        New Orleans       Orleans  \
1         Louisiana  New Orleans (Gentilly)        New Orleans       Orleans   
2         Louisiana   New Orleans (Algiers)        New Orleans       Orleans   
3         Louisiana       Saint Martinville  Saint Martinville    St. Martin   
4         Louisiana                Tallulah           Tallulah       Madison   
...             ...                     ...                ...           ...   
1158        Wyoming                Cheyenne           Cheyenne       Laramie   
1159          Idaho                Caldwell           Caldwell        Canyon   
1160          Idaho                   Boise              Boise           Ada   
1161  New Hampshire              Manchester         Manchester  Hillsborough   
1162         Hawaii                 Waianae            Waianae      Honolulu   

     State Code  
0            LA  
1            LA  
2            LA  
3            LA  
4            LA  
...         ...  
1158         WY  
1159         ID  
1160         ID  
1161         NH  
1162         HI  

[1163 rows x 5 columns]

In [144]:
search_counties_new_1 = search_counties_new_1.reindex(columns=['State','State Code','County','City','City Or County'], copy = True)

In [145]:
search_counties_new_1

State State Code        County               City   
0         Louisiana         LA       Orleans        New Orleans  \
1         Louisiana         LA       Orleans        New Orleans   
2         Louisiana         LA       Orleans        New Orleans   
3         Louisiana         LA    St. Martin  Saint Martinville   
4         Louisiana         LA       Madison           Tallulah   
...             ...        ...           ...                ...   
1158        Wyoming         WY       Laramie           Cheyenne   
1159          Idaho         ID        Canyon           Caldwell   
1160          Idaho         ID           Ada              Boise   
1161  New Hampshire         NH  Hillsborough         Manchester   
1162         Hawaii         HI      Honolulu            Waianae   

              City Or County  
0                New Orleans  
1     New Orleans (Gentilly)  
2      New Orleans (Algiers)  
3          Saint Martinville  
4                   Tallulah  
...                      ...  
1158                Cheyenne  
1159                Caldwell  
1160                   Boise  
1161              Manchester  
1162                 Waianae  

[1163 rows x 5 columns]

* Merging the required data with the gun_violence data to create a final data set for analysis

In [148]:
us_gunviolence_final = us_gunviolence.merge(right = search_counties_new_1, how = 'inner', left_on = ['State','City Or County'], right_on = ['State','City Or County'])

In [149]:
us_gunviolence_final

Incident ID       Incident Date       State      City Or County   
0          271363   December 29, 2014   Louisiana         New Orleans  \
1          191416  September 15, 2014   Louisiana         New Orleans   
2          175767     August 13, 2014   Louisiana         New Orleans   
3          172304     August 10, 2014   Louisiana         New Orleans   
4          152468       June 29, 2014   Louisiana         New Orleans   
...           ...                 ...         ...                 ...   
4216      2507836    January 21, 2023  California       Monterey Park   
4217      2503345    January 16, 2023  California              Goshen   
4218      2503649    January 16, 2023     Florida         Fort Pierce   
4219      2495648     January 4, 2023        Utah  Cedar City (Enoch)   
4220      2495059     January 4, 2023    Virginia            Dumfries   

                                       Address  # Killed  # Injured   
0                          Poydras and Bolivar         0          4  \
1         Chef Menteur Highway and Dale Street         0          5   
2     New Orleans and North Rocheblave streets         0          4   
3                    5400 block of Burgundy St         2          5   
4                      700 block of Bourbon St         1          9   
...                                        ...       ...        ...   
4216                          122 W Garvey Ave        12          9   
4217                     Harvest Ave and Rd 68         6          0   
4218                    Avenue M and N 13th St         1          7   
4219                          4923 N Albert Dr         8          0   
4220                  17900 block of Milroy Dr         1          4   

     State Code          County           City  
0            LA         Orleans    New Orleans  
1            LA         Orleans    New Orleans  
2            LA         Orleans    New Orleans  
3            LA         Orleans    New Orleans  
4            LA         Orleans    New Orleans  
...         ...             ...            ...  
4216         CA     Los Angeles  Monterey Park  
4217         CA          Tulare         Goshen  
4218         FL       St. Lucie    Fort Pierce  
4219         UT            Iron     Cedar City  
4220         VA  Prince William       Dumfries  

[4221 rows x 10 columns]

In [151]:
us_gunviolence_final = us_gunviolence_final.reindex(columns = ['Incident ID', 'Incident Date', 'State', 'State Code', 'County', 'City','City Or County','Address','# Killed','# Injured'], copy = True)

In [152]:
us_gunviolence_final

Incident ID       Incident Date       State State Code          County   
0          271363   December 29, 2014   Louisiana         LA         Orleans  \
1          191416  September 15, 2014   Louisiana         LA         Orleans   
2          175767     August 13, 2014   Louisiana         LA         Orleans   
3          172304     August 10, 2014   Louisiana         LA         Orleans   
4          152468       June 29, 2014   Louisiana         LA         Orleans   
...           ...                 ...         ...        ...             ...   
4216      2507836    January 21, 2023  California         CA     Los Angeles   
4217      2503345    January 16, 2023  California         CA          Tulare   
4218      2503649    January 16, 2023     Florida         FL       St. Lucie   
4219      2495648     January 4, 2023        Utah         UT            Iron   
4220      2495059     January 4, 2023    Virginia         VA  Prince William   

               City      City Or County   
0       New Orleans         New Orleans  \
1       New Orleans         New Orleans   
2       New Orleans         New Orleans   
3       New Orleans         New Orleans   
4       New Orleans         New Orleans   
...             ...                 ...   
4216  Monterey Park       Monterey Park   
4217         Goshen              Goshen   
4218    Fort Pierce         Fort Pierce   
4219     Cedar City  Cedar City (Enoch)   
4220       Dumfries            Dumfries   

                                       Address  # Killed  # Injured  
0                          Poydras and Bolivar         0          4  
1         Chef Menteur Highway and Dale Street         0          5  
2     New Orleans and North Rocheblave streets         0          4  
3                    5400 block of Burgundy St         2          5  
4                      700 block of Bourbon St         1          9  
...                                        ...       ...        ...  
4216                          122 W Garvey Ave        12          9  
4217                     Harvest Ave and Rd 68         6          0  
4218                    Avenue M and N 13th St         1          7  
4219                          4923 N Albert Dr         8          0  
4220                  17900 block of Milroy Dr         1          4  

[4221 rows x 10 columns]

## Datetime conversion

* Converting the dates that are in string to datetime type

In [159]:
us_gunviolence_final['Incident Date'] = pd.to_datetime(us_gunviolence_final['Incident Date'])

In [161]:
us_gunviolence_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4221 entries, 0 to 4220
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Incident ID     4221 non-null   int64         
 1   Incident Date   4221 non-null   datetime64[ns]
 2   State           4221 non-null   object        
 3   State Code      4221 non-null   object        
 4   County          4221 non-null   object        
 5   City            4221 non-null   object        
 6   City Or County  4221 non-null   object        
 7   Address         4209 non-null   object        
 8   # Killed        4221 non-null   int64         
 9   # Injured       4221 non-null   int64         
dtypes: datetime64[ns](1), int64(3), object(6)
memory usage: 329.9+ KB


In [167]:
us_gunviolence_final.head()

Incident ID Incident Date      State State Code   County         City   
0       271363    2014-12-29  Louisiana         LA  Orleans  New Orleans  \
1       191416    2014-09-15  Louisiana         LA  Orleans  New Orleans   
2       175767    2014-08-13  Louisiana         LA  Orleans  New Orleans   
3       172304    2014-08-10  Louisiana         LA  Orleans  New Orleans   
4       152468    2014-06-29  Louisiana         LA  Orleans  New Orleans   

  City Or County                                   Address  # Killed   
0    New Orleans                       Poydras and Bolivar         0  \
1    New Orleans      Chef Menteur Highway and Dale Street         0   
2    New Orleans  New Orleans and North Rocheblave streets         0   
3    New Orleans                 5400 block of Burgundy St         2   
4    New Orleans                   700 block of Bourbon St         1   

   # Injured  
0          4  
1          5  
2          4  
3          5  
4          9

## exporting the final dataset to .csv for further analysis in POWER BI

In [170]:
# exporting data to local storage

us_gunviolence_final.to_csv(r'D:\DA_Projects\Gun Violence USA\gun_violence_data.csv',index = False)

## spliting the data sets for power BI Consumption

In [1]:
import pandas as pd
import numpy as np


In [2]:
us_gunviolence_final = pd.read_csv(r'D:\DA_Projects\Gun Violence USA\gun_violence_data.csv')

In [8]:
# ensuring that the Incident IDs are unique

len(us_gunviolence_final['Incident ID']) == len(us_gunviolence_final['Incident ID'].unique())

True

In [9]:
us_gunviolence_final.head()

Incident ID Incident Date      State State Code   County         City   
0       271363    29-12-2014  Louisiana         LA  Orleans  New Orleans  \
1       191416    15-09-2014  Louisiana         LA  Orleans  New Orleans   
2       175767    13-08-2014  Louisiana         LA  Orleans  New Orleans   
3       172304    10-08-2014  Louisiana         LA  Orleans  New Orleans   
4       152468    29-06-2014  Louisiana         LA  Orleans  New Orleans   

  City Or County                                   Address  # Killed   
0    New Orleans                       Poydras and Bolivar         0  \
1    New Orleans      Chef Menteur Highway and Dale Street         0   
2    New Orleans  New Orleans and North Rocheblave streets         0   
3    New Orleans                 5400 block of Burgundy St         2   
4    New Orleans                   700 block of Bourbon St         1   

   # Injured  
0          4  
1          5  
2          4  
3          5  
4          9

In [12]:
us_gunviolence_place = us_gunviolence_final[['Incident ID','Incident Date','State','County','Address']].copy()

In [13]:
us_gunviolence_place.head()

Incident ID Incident Date      State   County   
0       271363    29-12-2014  Louisiana  Orleans  \
1       191416    15-09-2014  Louisiana  Orleans   
2       175767    13-08-2014  Louisiana  Orleans   
3       172304    10-08-2014  Louisiana  Orleans   
4       152468    29-06-2014  Louisiana  Orleans   

                                    Address  
0                       Poydras and Bolivar  
1      Chef Menteur Highway and Dale Street  
2  New Orleans and North Rocheblave streets  
3                 5400 block of Burgundy St  
4                   700 block of Bourbon St

In [14]:
us_gunviolence_killed = us_gunviolence_final[['Incident ID','Incident Date','# Killed']].copy()

In [15]:
us_gunviolence_killed.head()

Incident ID Incident Date  # Killed
0       271363    29-12-2014         0
1       191416    15-09-2014         0
2       175767    13-08-2014         0
3       172304    10-08-2014         2
4       152468    29-06-2014         1

In [16]:
us_gunviolence_injured = us_gunviolence_final[['Incident ID','Incident Date','# Injured']].copy()

In [17]:
us_gunviolence_injured.head()

Incident ID Incident Date  # Injured
0       271363    29-12-2014          4
1       191416    15-09-2014          5
2       175767    13-08-2014          4
3       172304    10-08-2014          5
4       152468    29-06-2014          9

In [19]:
us_gunviolence_city = us_gunviolence_final[['Incident ID','Incident Date','City','City Or County']].copy()

In [20]:
us_gunviolence_city.head()

Incident ID Incident Date         City City Or County
0       271363    29-12-2014  New Orleans    New Orleans
1       191416    15-09-2014  New Orleans    New Orleans
2       175767    13-08-2014  New Orleans    New Orleans
3       172304    10-08-2014  New Orleans    New Orleans
4       152468    29-06-2014  New Orleans    New Orleans

In [27]:
us_state_code = us_gunviolence_final[['State','State Code']].drop_duplicates().copy()

In [31]:
us_state_code.head()

State State Code
0     Louisiana         LA
75   California         CA
155    Illinois         IL
156    Missouri         MO
223    Kentucky         KY

## exporting all the split data sets

In [35]:
us_gunviolence_place.to_csv(r'D:\DA_Projects\Gun Violence USA\us_gunviolence_place.csv',index = False)
us_gunviolence_killed.to_csv(r'D:\DA_Projects\Gun Violence USA\us_gunviolence_killed.csv',index = False)
us_gunviolence_injured.to_csv(r'D:\DA_Projects\Gun Violence USA\us_gunviolence_injured.csv',index = False)
us_gunviolence_city.to_csv(r'D:\DA_Projects\Gun Violence USA\us_gunviolence_city.csv',index = False)
us_state_code.to_csv(r'D:\DA_Projects\Gun Violence USA\us_state_code.csv',index = False)